VMs/metrics/README.md

```diff
@@ -0,0 +1,242 @@
+This script is expected to be invoked either by Jenkins upon agent disconnect, or
+by hand if testing VMs.
+
+## Example
+
+```bash
+# Run this once ever normally, unless you need to renew your keys.
+$ ./1_create_keys.sh
+# Now start to build the master VM image.
+$ ./2_build_baseimage.sh ubuntu16.04
+# And wait for the cloud-init 'finished' line.
+# Now press ^]^]^] to get back to the host, and leave the master VM running.
+# And complete the master VM initialisation.
+$ ./3_complete_baseimage.sh ubuntu16.04
+# At this point you have a master VM image installed, which you can view with virsh:
+$ virsh list --all
+ Id    Name                           State
+----------------------------------------------------
+ -     ubnt16.04_master               shut off
+
+# If you wish to hand-debug a VM at this point, then you might also want to:
+$ ./4_clone_vm.sh ubuntu16.04
+$ ./5_login_clone.sh ubuntu16.04
+# You will now be SSH'd into the clone VM.
+$ ./7_delete_clone.sh ubuntu16.04
+
+# Otherwise, now you have a master VM set up you are ready to configure Jenkins.
+```
+
+## Configuring Jenkins
+
+To use these scripts we have to manipulate the Jenkins build node configuration somewhat.
+Ideally we could have used the [Jenkins Libvirt slaves plugin](https://plugins.jenkins.io/libvirt-slave)
+to manage our slave VMs, but upon testing, the 'revert' to the base clean VM function did
+not perform as expected.
+
+To that end, we manipulate the Jenkins 'ssh launch' pre/post command configuration to
+actually launch and remove our VMs.
+
+We also add a small manipulation to prevent Jenkins from trying to launch the agent.jar
+itself on the host system (as our scripts now handle that launch inside the VM).
+
+To try and ensure we get a fresh clean VM per-PR/build we also configure the nodes to be
+as short lived as we can. Again, we would have liked to have used something like the
+[Single Use Slave Plugin](https://plugins.jenkins.io/singleuseslave) - but, that really
+does provide a single use slave node, which will then not launch again.
+
+Whilst our configuration ends up not perfect (the shortest time we seem to be able to set
+to kill off an inactive node is 1 minute, so if another job is in the queue or arrives
+within 1 minute of a previous job, then it will be scheduled to build on the same VM instance), the expectation is that we will have enough gaps that we acquire enough new VMs to vastly
+reduce the 'dirty node' instabilities we saw with pure bare-metal builds.
+
+Also, in the event of getting a dirty VM node, a simple 'node offline/online' in the
+Jenkins UI should perform a VM shutdown/clone/run cycle, and thus get us back to having
+a stable clean VM on that node.
+
+Below are examples of the configuraiton dialogs:
+
+### Setup the pre-post scripts
+
+On your slave VM node host machine, create a pair of scripts specific to your host to allow
+Jenkins to invoke the sub-scripts. For example:
+
+start_gitvm.sh
```

Comment:
Might be worth putting this script name (and the one below) in backticks to make it stand out as a command.